File: 0224-basic-calculator/0224-basic-calculator.java
class Solution {
    public int calculate(String s) {
     // Stack to hold the intermediate results and signs
        Deque<Integer> stack = new ArrayDeque<>();
        // Initialize the sign as positive
        int sign = 1;
        // This will hold the final result of the evaluation
        int result = 0;
        // Length of the input string for iteration
        int length = s.length();

        // Iterating over each character in the string
        for (int i = 0; i < length; ++i) {
            char ch = s.charAt(i);
            // Check if the current char is a digit
            if (Character.isDigit(ch)) {
                int startIndex = i;
                int number = 0;
                // Build the number till we encounter a non-digit
                while (startIndex < length && Character.isDigit(s.charAt(startIndex))) {
                    number = number * 10 + s.charAt(startIndex) - '0';
                    startIndex++;
                }
                // Update the result with the current number and sign
                result += sign * number;
                // Move the pointer to the end of the number
                i = startIndex - 1;
            } else if (ch == '+') {
                // Set sign as positive for addition
                sign = 1;
            } else if (ch == '-') {
                // Set sign as negative for subtraction
                sign = -1;
            } else if (ch == '(') {
                // Push the result and sign on the stack before the parenthesis
                stack.push(result);
                stack.push(sign);
                // Reset result and sign for the expression inside the parenthesis
                result = 0;
                sign = 1;
            } else if (ch == ')') {
                // After closing parenthesis, result is the evaluated value inside the parenthesis
                // Pop the sign before the parenthesis and multiply it with the current result,
                // then add the result before the parenthesis
                result = stack.pop() * result + stack.pop();
            }
        }
        // Return the final result of the evaluation
        return result;
    }
}

File: 0224-basic-calculator/0224-basic-calculator.cpp
class Solution {
public:
    int calculate(string s) {
        stack<int> operands;  // Stack to hold numbers and signs (as integer)
        int result = 0;       // Accumulator for the current result
        int sign = 1;         // Current sign, starts with positive
        int n = s.size();     // Size of the input string
      
        for (int i = 0; i < n; ++i) {
            // If the character is a digit, it could be part of a multi-digit number.
            if (isdigit(s[i])) {
                int num = 0; // Initialize the number to be 0
                // Construct the number from the subsequent digits
                while (i < n && isdigit(s[i])) {
                    num = num * 10 + (s[i] - '0');
                    ++i;
                }
                // Add/Subtract the number to/from the current result
                result += sign * num;
                // Since the loop will increment i after, we decrease it here
                --i;
            } else if (s[i] == '+') {
                sign = 1;  // Set sign to positive for the next number
            } else if (s[i] == '-') {
                sign = -1; // Set sign to negative for the next number
            } else if (s[i] == '(') {
                // When encountering '(', push the current result and sign onto the stack
                operands.push(result);
                operands.push(sign);
                // Reset result and sign for the new sub-expression inside the parentheses
                result = 0;
                sign = 1;
            } else if (s[i] == ')') {
                // Ending a sub-expression, pop sign from stack and multiply it with the current result
                result *= operands.top();
                operands.pop();
                // Add the result of the sub-expression to the result up to the '('
                result += operands.top();
                operands.pop();
            }
            // Ignore whitespace and other non-math characters
        }
        return result; // Return the final result
    }
};